[Doc] [C++20] [Modules] Clarify the reachability of internal partition units (#102572)
llvm/llvm-project: clang/docs/StandardCPlusPlusModules.rst

imported modules don't provide such a header, one can be made manually for
improved compile time performance.

Reachability of internal partition units
----------------------------------------

The internal partition units are called as implementation partition unit somewhere else.
But the name may be confusing since implementation partition units are not implementation

> units.

Author: done

> According to [module.reach]p1,2:

Author: It was from the draft version (https://eel.is/c++draft/module.reach#1). I changed it to the links. Hope this helps.

> According to [module.reach]p1,2:

Reviewer: Links only sort of help -- because they're live links, they get out of date very quickly. Putting a standard version number there helps the reader to understand which standard the quote was true for and they can hunt for the wording in the latest draft from there.

  A translation unit U is necessarily reachable from a point P if U is a module
  interface unit on which the translation unit containing P has an interface
  dependency, or the translation unit containing P imports U, in either case
  prior to P.

  All translation units that are necessarily reachable are reachable. Additional
  translation units on which the point within the program has an interface
  dependency may be considered reachable, but it is unspecified which are and
  under what circumstances.

For example,

.. code-block:: c++

  // a.cpp
  import B;
  int main()
  {
      g<void>();
  }

  // b.cppm
  export module B;
  import :C;
  export template <typename T> inline void g() noexcept
  {
      return f<T>();
  }

  // c.cppm
  module B:C;
  template<typename> inline void f() noexcept {}

> The internal partition units ``c.cppm`` is not necessarily reachable to
> ``a.cpp`` since ``c.cppm`` is not a module interface unit and ``a.cpp``
> doesn't import ``c.cppm``. Then it is up to the compiler to decide if
> ``c.cppm`` is reachable to ``a.cpp`` or not. Clang's decision is the
> non-directly imported internal partition units are not reachable.

Reviewer: ```suggestion
The internal partition unit ``c.cppm`` is not necessarily reachable by
``a.cpp`` because ``c.cppm`` is not a module interface unit and ``a.cpp``
doesn't import ``c.cppm``. This leaves it up to the compiler to decide if
``c.cppm`` is reachable by ``a.cpp`` or not. Clang's behavior is that
indirectly imported internal partition units are not reachable.
```

Author: Done

> The suggestion to use internal partition units is, only import them in
> the implementation units unless we understand the codebases very well.

Reviewer: ```suggestion
The suggested approach for using an internal partition unit in Clang is to only import them in
the implementation unit.
```

Author: Done
